source/_posts/2017-08-11-How-we-debug-with-Charles.md

```diff
@@ -0,0 +1,103 @@
+
+1. Monitor all traffic between API and application (Recording).
+2. Simulating network conditions (Throttle).
+3. Rewriting the API (Rewrite).
+4. Simulating API responses (Map local).
+5. Interrupting API calls (Breakpoints).
+
+### 1. Monitor all traffic
+
+The most common thing you can do with Charles, is to have a peek underneath the hood of the application, also known as 'Recording'. By default in Charles, when you start to record traffic you will be looking at the structure tab, that lists all the executed requests organised by URL. All traffic in Charles is stored in the current session, which acts as a log over all traffic. Each request appears in the session window when they are recorded. The session allows testers to look into all traffic coming from the app: from opening the app for the first time, to creating users, to logging in or logging out. We make sure to keep Charles running in the background while performing all our tests to the application. If the app suddenly fails during testing, we can easily 'go back in time' and investigate why, by looking at the requests in the session. Here are a few scenarios in which Charles is a great help in debugging the application:
+
+* When the app is making network requests to an API or a third party provider and you are not receiving the expected results in the app.
+* The app displays an empty view, when the app is supposed to fetch data from the API – you need to make sure to check whether data is received correctly from the API.
+* If you are making changes in the app and the changes are not applied – you need to check if the app is sending the data correctly or whether you are receiving the processed data from the server unexpectedly.
+* Any other scenario that involves the app making or receiving network calls – you need to check your Charles!
+
+With Charles, we are not only able to monitor the implementation of API, we can also monitor the implementation of third party providers such as push providers, Google Analytics, Hockey and furthermore we can monitor the order in which they are implemented. To ensure an agile workflow and a quick response to the issues that are found. Charles sessions can be exported and shared across the team. This 'comes in handy' when a tester is discovering an issue and wants to show the developer what is happening without the aforementioned steps to reproduce the issue. By exporting the session from Charles and adding it as an attachment in the ticket, it becomes very easy for the developer to reproduce the problem. Charles also has a built in way to publish gists to Github. The gist will automatically contain all required information in order to reproduce the issue such as auth token, request headers, API path, API environment, etc, etc.. This makes the workflow around discovering and reporting bugs in API or API implementation very quick and agile.
+
```

Comment:
Generally, you have a lot text in each paragraphs which seems to be a bit hard to focus on. Maybe split the big chunks of text into smaller ones? Or add some extra styling to it.

Reply:
> To ensure an agile workflow and a quick response to the issues that are found. Charles sessions can be exported and shared across the team.

Those 2 phrases should be combined into one. So have "...found, Charles..." You could also switch to a new paragraph before this.

Reply:
>  Charles also has a built in way to publish gists to Github. 

New paragraph before this?

Reply:
> such as auth token, request headers, API path, API environment, etc, etc..

One `etc.` should be sufficient.